Updates to README.md to help with readability

gilcrest opened the pull request:
Hey - I've recently decided to switch to zerolog and wanted to try and help by adding some organization to the README.md file.  I love the logger so far!  As I'm teaching myself the library and learning the APIs, it seems like there's some room for improvement in the instructions for the layman (that would be me)...  I have more edits I'd like to make as I go, but wanted to see if you considered this helpful before I spent more time on it!  Thanks for the logger and your consideration!

You can see the changes fully realized at my fork, which may be easier to read than all the code merge: [https://github.com/gilcrest/zerolog](https://github.com/gilcrest/zerolog)

Dan

rs commented on README.md:
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
-### A global logger can be use for simple logging
+import (
+	"github.com/rs/zerolog/log"
+)
```
You might make this one-line to reduce the size of the example and focus on the important stuff.

gilcrest replied:
No problem

rs commented on README.md:
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+- info (InfoLevel, 1)
+- debug (DebugLevel, 0)
 
+You can set the Global logging level to any of these options using the SetGlobalLevel function in the zerolog package, passing in one of the given constants above, e.g. zerolog.InfoLevel would be the "info" level.  Whichever level is chosen, all logs with a level greater than or equal to that level will be written. To turn off logging entirely, pass the Disabled constant.
```
Could you please backquote "code" like (`SetGlobalLevel` and `zerolog.InfoLevel`) in the text?

gilcrest replied:
Yes, will do

rs commented on README.md:
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+
+func main() {
+	//zerolog.SetGlobalLevel(zerolog.InfoLevel) // Run 1
+	zerolog.SetGlobalLevel(zerolog.DebugLevel) // Run 2
```
Would be nicer with the use of a `-d` flag instead.

gilcrest replied:
OK, this is a good excuse to teach myself flags...  Give me a little time on this one

gilcrest commented on README.md:
````diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
 ```
+## Getting Started
+### Simple Logging
+For simple logging, import the global logger package **github.co/rs/zerolog/log**
````
Yup, will do

gilcrest commented on README.md:
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
-log.Print("hello world")
+func main() {
+	// I like UNIX Time - if you set zerolog.TimeFieldFormat to an
+	// empty string, logs will write with UNIX time
```
Good to know!  I'll change this slightly to reflect that as well...

rs commented:
Awesome, I'll merge where you're done.

What about the `log_example_test.go`?

gilcrest commented:
Oh right, I'll add the examples to that as well

gilcrest commented:
Hrm, so I started to add the examples from the README into `log_example_test.go`, but by nature, they all log with the current time, which is always variable and cannot be predicted in a static output...  What do you think if I add something like this to the package - it will pass the test as there's no "Output"...  Unless you have a better idea?

```go
package zerolog_test

import (
	"errors"
	stdlog "log"
	"os"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
)

// The below is a simple logging example using the log package.
// Note that "Output" is renamed to "Produces" as otherwise, this
// test will not pass. Log.Print has time in its Output, which
// cannot be predicted in a static Output
func ExampleSimple() {
	// UNIX Time is faster and smaller than most timestamps
	// If you set zerolog.TimeFieldFormat to an empty string,
	// logs will write with UNIX time
	zerolog.TimeFieldFormat = ""

	log.Print("hello world")

	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}

}
```

gilcrest commented:
Also, one thing - having to add the `zerolog.TimeFieldFormat = ""` is obviously adding length to these examples and an extra import...  When I installed the package, by default I was getting "normal" timestamp ("time":"2018-01-17T10:46:18-05:00") - I can shorten all these if you're ok with displaying this type of timestamp in the various examples, but you had been using UNIX time in all your examples, so I stuck with that...  Might it even be better to have UNIX time as the default though?

rs commented:
In test mode we could make the timestamp static.

Use `zerolog.TimeFieldFormat = ""` would be the obvious choice but I don't want to break people log pipeline with such change.

gilcrest commented:
I added 3 of the 4 examples -- I put in the expected `Output:` - unfortunately, I'm not sure what you mean by "In test mode we could make the timestamp static."  I removed Output for now in order to get them in there and the tests don't fail.  I tried a few different things locally, but at a bit of a loss - sorry - learning Go n00b-ness.  Also, for the example where you say "Would be nicer with the use of a -d flag instead."  I have no idea what a -d flag is... I searched through the flags doc a bunch of blogs, etc but no mention of a -d flag... If you can help point me in the right direction, I should be able to figure it out...

rs commented:
Basically you isolate the part that set the timestamp to "now" into a global function variable (like `timeNowFunc`) and in the _test.go's `init` function, you override this function to return a fixed time. This way, all tests will get a constant timestamp.

By `-d` flag, I mean creating a `-debug` or `-d` flag that would change the log level to debug (the default being set to info for instance. This would let you modulate the log level with a cli parameter instead of having to recompile with different commented lines.

gilcrest commented:
ok, super helpful comments, this is great - I'm learning - thank you - getting there... sorry for all the back and forth...  I have just checked in the changes for the Global Log Level section using flags - it's slightly different than what you were saying - can you let me know if you're ok with the way I did it?

rs commented on README.md:
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+	"github.com/rs/zerolog/log"
+)
+
+var lvl string
```
No need for global variable if you move the flag into the main.

gilcrest replied:
ok got it - thx

rs commented on README.md:
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+
+// Declare "level" flag and bind it to lvl variable
+func init() {
+	flag.StringVar(&lvl, "level", "error", "level is the logging level (debug, info, warn, etc.")
```
You should move this to the main. It's not a great pattern to have your flag as globals.

You may also use this notation:

```go
lvl := flag.String("level", …)
```

I would stick with a `debug` flag to simplify the example, even if you're right, this approach is more realistic.

gilcrest replied:
Understood - thanks again!

rs commented on README.md:
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+	// Set the logging level based on the command line flag
+	zerolog.SetGlobalLevel(level(lvl))
+
+	log.Debug().Msg("filtered out message")
```
The messages are no longer relevant.

gilcrest replied:
Updated messages to hopefully be clearer...

rs commented on log_example_test.go:
```diff
@@ -394,3 +421,19 @@ func ExampleContext_Durs() {
+
+	// Produces: {"time":1516133263,"level":"fatal","error":"A repo man spends his life getting into tense situations","service":"myservice","message":"Cannot start myservice"}
+	//         exit status 1
+
```
Unneeded `exit status 1` and extra blank line.

gilcrest replied:
dropped

rs commented on log_example_test.go:
```diff
@@ -7,8 +7,35 @@ import (
+	log.Info().Msg("hello world")
+
+	// Produces: {"time":1516227604,"level":"info","message":"hello world"}
+
```
Extra blank line.

gilcrest replied:
Purged

rs commented on log_example_test.go:
```diff
@@ -7,8 +7,35 @@ import (
+	log.Print("hello world")
+
+	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}
+
```
Extra blank line.

gilcrest replied:
Gone

rs commented on README.md:
```diff
@@ -24,43 +24,150 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
+		lvl = zerolog.InfoLevel
+	}
+
+	zerolog.SetGlobalLevel(lvl)
```
Cosmetic comment: I would have written this that way to better show what's default:

```go
zerolog.SetGlobalLevel(zerolog.InfoLevel)
if *debug {
    zerolog.SetGlobalLevel(zerolog.DebugLevel)
}
```

gilcrest replied:
You're totally right - changed and committed!

rs commented on log_example_test.go:
```diff
@@ -7,8 +7,33 @@ import (
+
+	log.Print("hello world")
+
+	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}
```
Are you going to give the static test time trick a try?

gilcrest replied:
Yep!

gilcrest replied:
I'll try and figure it out ASAP...  I need to move on to learn the rest of your APIs as well - let me know if this is helpful and I can continue to help fill out some of the documentation.

gilcrest commented:
Here's the latest - I moved the examples out of log_example_test.go in the zerolog package and into a new file (with the same name) in the `zerolog/log` package since those are the functions I'm truly writing examples for... let me know if you think that makes sense.  

I'm at a bit of a loss now though...  I isolated the time like you recommended and set it to a static time and the function is now providing the correct static output, however, the examples aren't working.  I tried overriding the global logger to ensure the output is being directed to the stdout, but still no luck...  See below - the output is clearly written, but for some reason, the test isn't picking it up (`got:` is blank?).  Everything works fine when I create a logger from the regular `zerolog` package, it's when using the `zerolog/log` package where I hit this issue...

```bash
{"time":1199811905,"level":"debug","message":"hello world"}
--- FAIL: ExamplePrint (0.00s)
got:

want:
{"time":1199811905,"level":"debug","message":"hello world"}
FAIL
```

rs commented:
There is a strange issue with the testing framework. If you put `log.Logger = zerolog.New(os.Stdout).With().Timestamp().Logger()` at the beginning of each example, it fixes the issue. It does not work if you put it in the init function. I'm not sure why.

gilcrest commented:
Oh man - thanks for pointing that out, I was pulling my hair out trying to figure out what the hell I was doing wrong.  Submitted another commit, this time with more examples - let me know if this looks ok or if you need anything else.

rs commented:
@gilcrest do you need some help to complete this PR?

gilcrest commented:
Hey @rs - I wasn't sure if you're ok with everything at this point?  Do I need to do anything else to finalize the PR?

rs commented:
Fair enough
